### ISSUE TITLE
libc libraries get included in the bundle

### ISSUE THREAD

**Issue by EfremBraun** [pyinstaller/pyinstaller#6993]
## Description of the issue
libstdc++ and other libc libraries get included in the bundle I make using pyinstaller, despite https://pyinstaller.org/en/stable/usage.html#making-gnu-linux-apps-forward-compatible stating that it will not. When I build my package on Fedora 35 and then try to run it on Fedora 36, this results in the error:
```
libGL error: MESA-LOADER: failed to open swrast: /home/efremdan1/Documents/MyPackage/libstdc++.so.6: version `GLIBCXX_3.4.30' not found (required by /usr/lib64/dri/swrast_dri.so) (search paths /usr/lib64/dri, suffix _dri)
libGL error: failed to load driver: swrast
```

If I simply remove the libstdc++ file from the bundle, the bundle runs without any error since it uses what's found on the user's computer.

This is actually quite an old issue. In the helpful files https://github.com/mherrmann/fbs/blob/master/fbs/freeze/linux.py, https://github.com/mherrmann/fbs/blob/master/fbs/freeze/ubuntu.py, and https://github.com/mherrmann/fbs/blob/master/fbs/freeze/fedora.py, it was recommended to remedy this issue by simply removing these unnecessary files. However, I feel like this is something that should be done by pyinstaller itself, especially given the information in the pyinstaller docs.

### Context information (for bug reports)

* Output of `pyinstaller --version`: 5.2
* Version of Python: 3.9

**Comment by rokm** [comment]
While our docs often contain questionable information, I'd say that the linked section is specifically referring to the standard *C* library, `libc`, and not to the standard C++ library, `libstdc++`.

So the described behavior is correct for "simple" applications with dependencies that are in turn dependent only on the standard C library. As soon as you start bundling complex stuff (like UI libraries) that (directly or indirectly) depend on libstdc++, you've entered the gray area of ABI compatibility lottery on linux.

EDIT: or rather, to turn things around, the recommendation to "build on oldest version of linux you want to support" stems from the fact that the python shared library collected from the system depends on version of libc from the build system (or later, due to glibc backwards compatibility), and will refuse to load on systems with earlier version of the libc.

> If I simply remove the libstdc++ file from the bundle, the bundle runs without any error since it uses what's found on the user's computer.

Yeah, but this is situation-dependent, and there's no one-site-fits-all rule regarding the (non-)collected shared libraries.

Had you run the bundle with removed `libstdc++` on a distribution of compatible vintage (in terms of libc) but with slightly older `libstdc++`, the collected shared libs that depend on symbols from original (newer) `libstdc++` would fail to load with missing versioned symbols. Of course, you get away with removing those collected shared libs as well, but that again depends on what is available on the user's system.

> This is actually quite an old issue.

It is, but it is a problem of binary-only software on linux in general (for example, in my university days, I always had to remove bundled libstdc++ et al. from the MATLAB installation to make it run on my bleeding-edge Fedora box). It is the reason why you generally don't use binary packages from one distribution or another (sometimes not even across different versions of the same distribution), and it is the reason why common runtime environments, like that provided by flatpak, are a thing.

> it was recommended to remedy this issue by simply removing these unnecessary files. However, I feel like this is something that should be done by pyinstaller itself, especially given the information in the pyinstaller docs.

I'd argue that you are in much better position to judge whether the files are unnecessary or not, since you know what distribution(s) you are supposed to be targeting. PyInstaller cannot know that. 

So I'll be strongly pushing for status quo on this one, because changing the behavior will simply trade one sort of the problem for another one. And I think with current PyInstaller design, it is generally easier to remove shared libs from the bundle that adding them.

**Comment by EfremBraun** [comment]
Sounds like a very reasonable point-of-view. Might I suggest that the docs be updated with some of this information to help future developers?

I'd recommend adding to the bottom of the aforementioned section something along the lines of the following paragraph:
> PyInstaller does bundle the C++ standard library (libstdc++) with the app. If your app depends on this library, including it is useful for distribution of the app to systems with older versions of the C++ standard library, but including it can cause errors when distributed to systems with newer versions of the C++ standard library. Depending on the distributions you are targeting, you may want to manually remove this library after the PyInstaller bundling process finishes.

I can put in a pull request with this added into the docs if you agree.

The only thing is that it might be a better idea to be a bit more general. Per the fbs links I included above, other libraries may need to be removed as well (e.g., libtinfo, libreadline, libdrm, libgpg-error, libgtk, libgio). I'm not sure what the deal is what them though, so I don't know how to write them into the paragraph...

**Comment by rokm** [comment]
> Might I suggest that the docs be updated with some of this information to help future developers?

Yeah, that sounds like a good idea.

But like you mentioned, this should be written in a more general way, as it applies to pretty much all shared libraries except the standard C library. Something along the lines of:

>Note that PyInstaller does bundle other shared libraries that are discovered via dependency analysis, such as `libstdc++.so.6`, `libfontconfig.so.1`, `libfreetype.so.6`. These libraries may be required on systems where older (and thus incompatible) versions of these libraries are available. On the other hand, the bundled libraries may cause issues when trying to load a system-provided shared library that is linked against a newer version of (system-provided) library. 
>
>For example, system-installed mesa DRI drivers (e.g., `radeonsi_dri.so`) depend on the system-provided version of  `libstdc++.so.6`. If the frozen application bundles an older version of `libstdc++.so.6` (as collected from the build system), this will likely cause missing symbol errors and prevent the DRI drivers from loading. In this case, the bundled `libstdc++.so.6` should be removed. However, this may not work on a different distribution that provides `libstdc++.so.6` older than the one from the build system; in that case, the bundled version should be kept, because the system-provided version may lack the symbols required by other collected binaries that depend on `libstdc++.so.6`.

**Comment by EfremBraun** [comment]
Putting in the pull request now...

**Comment by bwoodsend** [comment]
I'd have thought though that the idea of building on old docker images is that the versions of the libraries you intend to link against dynamically are lower than or equivalent to those of your intentionally supported runtime environments. In which case you're unlikely to encounter a libc++ at runtime which is older than that of the build image. So the scenario where including libc++ is the correct move is pretty rare, right? I'd have also thought that in such a case where the runtime libc++ really is older than the build-time libc++, libc is also likely to be older in the runtime environment in which case the application+runtime environment combination is already doomed?

So tl;dr, if you're going down the *build on old docker image* route, you should unconditionally leave libc++ behind?

**Comment by rokm** [comment]
> So tl;dr, if you're going down the _build on old docker image_ route, you should unconditionally leave libc++ behind?

I suppose in _build in old docker image route_, it would be reasonable to leave `libstdc++` out, yeah (unless you are targeting a very minimal environment that does not provide its copy...).

**Comment by bwoodsend** [comment]
Would mirroring manylinux would be a good default policy? i.e. Any of the libraries listed [here](https://peps.python.org/pep-0513/#the-manylinux1-policy) should be excluded? (Possibly we'd chuck openssl in there too.) I'd have thought that desktop applications are a much more likely PyInstaller use case than little CLI tools which can be ran in minimal docker images.

**Comment by rokm** [comment]
> Would mirroring manylinux would be a good default policy? i.e. Any of the libraries listed [here](https://peps.python.org/pep-0513/#the-manylinux1-policy) should be excluded? (Possibly we'd chuck openssl in there too.)

Hmm, perhaps. I'm still not fully convinced that we won't be just trading one set of problems for another, but I suppose it should, at least in theory, improve the situation for people who do adhere to the *build in old docker image* recommendation. I guess the only real answer is to make the change and see, if we are ready to deal with potential fallout. 

Although now that I think about it, it should be possible to manipulate the inclusion/exclusion lists at the top of the spec file, as a hacky way to bring any of the excluded libs back, if necessary...

**Comment by EfremBraun** [comment]
Given the discussion, should we open an issue about removing libstdc++ and other libraries from the distribution lest the issue be forgotten?

**Comment by bwoodsend** [comment]
I'll just reopen this one since it's a) not very long and b) contains the rational.

**Comment by omars44** [comment]
I'm struggling with this problem trying to build a flatpak  from a pyinstaller binary file, how do I exclude libstdc++, so that the mesa driver works autonomously with what it has

**Comment by rokm** [comment]
> I'm struggling with this problem trying to build a flatpak from a pyinstaller binary file, how do I exclude libstdc++, so that the mesa driver works autonomously with what it has

You need to build using the .spec file. In the spec file, filter `a.binaries` after `Analysis` is instantiated (and before its contents are passed to `PYZ`, `PKG`, `COLLECT`) to exclude the shared libraries. For flatpak, you might want to exclude all libraries from system library directories, which can be achieved by calling [`a.exclude_system_libraries`](https://github.com/pyinstaller/pyinstaller/blob/556cd59e66a8b034d8061267ef9b5dea05526821/PyInstaller/building/build_main.py#L798-L807) that's documented [here](https://pyinstaller.org/en/stable/spec-files.html#posix-specific-options).
